File: src/main/java/com/mashang/aicode/web/controller/AppController.java
package com.mashang.aicode.web.controller;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONUtil;
import com.mashang.aicode.web.ai.factory.AiCodeGenTypeRoutingServiceFactory;
import com.mashang.aicode.web.ai.factory.AppNameServiceFactory;
import com.mashang.aicode.web.ai.service.AiCodeGenTypeRoutingService;
import com.mashang.aicode.web.ai.service.AppNameService;
import com.mashang.aicode.web.ai.model.enums.CodeGenTypeEnum;
import com.mashang.aicode.web.model.enums.AppTypeEnum;
import com.mashang.aicode.web.annotation.AuthCheck;
import com.mashang.aicode.web.common.BaseResponse;
import com.mashang.aicode.web.common.DeleteRequest;
import com.mashang.aicode.web.common.ResultUtils;
import com.mashang.aicode.web.constant.AppConstant;
import com.mashang.aicode.web.constant.UserConstant;
import com.mashang.aicode.web.exception.BusinessException;
import com.mashang.aicode.web.exception.ErrorCode;
import com.mashang.aicode.web.exception.ThrowUtils;
import com.mashang.aicode.web.model.dto.app.AppAddRequest;
import com.mashang.aicode.web.model.dto.app.AppDeployRequest;
import com.mashang.aicode.web.model.dto.app.AppQueryRequest;
import com.mashang.aicode.web.model.dto.app.AppUpdateRequest;
import com.mashang.aicode.web.model.entity.App;
import com.mashang.aicode.web.model.entity.User;
import com.mashang.aicode.web.model.enums.AppTypeEnum;
import com.mashang.aicode.web.model.vo.AppVO;
import com.mashang.aicode.web.model.enums.PresetPromptEnum;
import com.mashang.aicode.web.model.vo.AppTypeVO;
import com.mashang.aicode.web.model.vo.PresetPromptVO;
import com.mashang.aicode.web.ratelimiter.annotation.RateLimit;
import com.mashang.aicode.web.ratelimiter.enums.RateLimitType;
import com.mashang.aicode.web.service.*;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.mashang.aicode.web.utils.IpUtils;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.http.MediaType;
import org.springframework.http.codec.ServerSentEvent;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.io.File;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * 应用 控制层。
 */
@RestController
@RequestMapping("/app")
@Slf4j
public class AppController {

    @Resource
    private AppService appService;

    @Resource
    private UserService userService;

    @Resource
    private ChatHistoryService chatHistoryService;

    @Resource
    private ProjectDownloadService projectDownloadService;
    @Resource
    private AiCodeGenTypeRoutingService aiCodeGenTypeRoutingService;
    @Resource
    private AiCodeGenTypeRoutingServiceFactory aiCodeGenTypeRoutingServiceFactory;
    @Resource
    private AppNameServiceFactory appNameServiceFactory;
    @Resource
    private SpaceUserService spaceUserService;

    @Resource
    private IpUtils ipUtils;


    /**
     * 下载项目
     *
     * @param appId
     * @param request
     * @param response
     */
    @GetMapping("/download/{appId}")
    public void downloadAppCode(@PathVariable Long appId, HttpServletRequest request, HttpServletResponse response) {

        ThrowUtils.throwIf(appId == null || appId <= 0, ErrorCode.PARAMS_ERROR, "应用ID无效");
        App app = appService.getById(appId);
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR, "应用不存在");
        User loginUser = userService.getLoginUser(request);
        if (!app.getUserId().equals(loginUser.getId())) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "无权限下载该应用代码");
        }
        String codeGenType = app.getCodeGenType();
        String sourceDirName = codeGenType + "_" + appId;
        String sourceDirPath = AppConstant.CODE_OUTPUT_ROOT_DIR + File.separator + sourceDirName;
        File sourceDir = new File(sourceDirPath);
        ThrowUtils.throwIf(!sourceDir.exists() || !sourceDir.isDirectory(), ErrorCode.NOT_FOUND_ERROR, "应用代码不存在，请先生成代码");
        String downloadFileName = String.valueOf(appId);
        projectDownloadService.downloadProjectAsZip(sourceDirPath, downloadFileName, response);
    }


    /**
     * 【用户】创建应用
     */
    @PostMapping("/add")
    public BaseResponse<Long> addApp(@RequestBody AppAddRequest appAddRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(appAddRequest == null, ErrorCode.PARAMS_ERROR);

        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        String initPrompt = appAddRequest.getInitPrompt();
        ThrowUtils.throwIf(StrUtil.isBlank(initPrompt), ErrorCode.PARAMS_ERROR, "初始化提示词不能为空");

        // 名称，使用 AI 自动生成
        AppNameService appNameService = appNameServiceFactory.createAppNameService();
        String appName = appNameService.generateAppName(initPrompt);
        log.info("AI 自动生成应用名称: {}", appName);

        // 创建应用
        App app = new App();
        BeanUtil.copyProperties(appAddRequest, app);
        app.setUserId(loginUser.getId());
        app.setAppName(appName);
        app.setSpaceId(appAddRequest.getSpaceId());
//        app.setPriority(0); // 默认非精选
        app.setPriority(0); // 默认优先级
        //ai自动选择应用类型
        AppTypeEnum appTypeEnum = aiCodeGenTypeRoutingService.routeAppType(appAddRequest.getAppName(), initPrompt);
        app.setAppType(appTypeEnum.getText());
        //ai自动选择生成模式
        AiCodeGenTypeRoutingService routingService = aiCodeGenTypeRoutingServiceFactory.createAiCodeGenTypeRoutingService();
        CodeGenTypeEnum selectedCodeGenType = routingService.routeCodeGenType(initPrompt);
        app.setCodeGenType(selectedCodeGenType.getValue());

        app.setIsDelete(0);

        boolean result = appService.save(app);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(app.getId());
    }

    /**
     * 【用户】根据 id 修改自己的应用（目前只支持修改应用名称）
     */
    @PostMapping("/update/user")
    public BaseResponse<Boolean> updateApp(@RequestBody AppUpdateRequest appUpdateRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(appUpdateRequest == null || appUpdateRequest.getId() == null, ErrorCode.PARAMS_ERROR);

        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        // 查询应用
        App app = appService.getById(appUpdateRequest.getId());
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR);

        // 校验权限（只能修改自己的应用）
        ThrowUtils.throwIf(!app.getUserId().equals(loginUser.getId()), ErrorCode.NO_AUTH_ERROR);

        // 更新应用名称
        App updateApp = new App();
        updateApp.setId(appUpdateRequest.getId());
        updateApp.setAppName(appUpdateRequest.getAppName());
        updateApp.setEditTime(new Date());

        boolean result = appService.updateById(updateApp);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(true);
    }

    /**
     * 【用户】根据 id 删除自己的应用（同时删除关联的对话历史）
     */
    @PostMapping("/delete/user")
    public BaseResponse<Boolean> deleteApp(@RequestBody DeleteRequest deleteRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(deleteRequest == null || deleteRequest.getId() == null, ErrorCode.PARAMS_ERROR);

        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        // 删除应用（同时删除关联的对话历史）
        boolean result = appService.deleteApp(deleteRequest.getId(), loginUser);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(true);
    }

    /**
     * 【用户】根据 id 查看应用详情
     */
    @GetMapping("/get/vo")
    public BaseResponse<AppVO> getAppVOById(long id, HttpServletRequest request) {
        ThrowUtils.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);

        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        // 查询应用
        App app = appService.getById(id);
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR);

        // 校验权限（只能查看自己的应用或精选应用）
        if (!app.getUserId().equals(loginUser.getId()) && app.getPriority() != 1) {
            throw new BusinessException(ErrorCode.NO_AUTH_ERROR);
        }
        // 浏览量增加
        App updateApp = new App();
        updateApp.setId(app.getId());
        updateApp.setPageViews(app.getPageViews() != null ? app.getPageViews() + 1 : 1L);
        appService.updateById(updateApp);


        AppVO appVO = appService.getAppVO(app);
        return ResultUtils.success(appVO);
    }

    /**
     * 【用户】分页查询自己的应用列表（支持根据名称查询，每页最多 20 个）
     */
    @PostMapping("/my/list/page/vo")
    public BaseResponse<Page<AppVO>> listMyAppVOByPage(@RequestBody AppQueryRequest appQueryRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(appQueryRequest == null, ErrorCode.PARAMS_ERROR);

        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        Page<AppVO> appVOPage = appService.listAppVOByPageForUser(appQueryRequest, loginUser.getId());
        return ResultUtils.success(appVOPage);
    }

    /**
     * 【用户】分页查询精选的应用列表（支持根据名称查询，每页最多 20 个）
     */
    @PostMapping("/featured/list/page/vo")
    public BaseResponse<Page<AppVO>> listFeaturedAppVOByPage(@RequestBody AppQueryRequest appQueryRequest) {
        ThrowUtils.throwIf(appQueryRequest == null, ErrorCode.PARAMS_ERROR);

        Page<AppVO> appVOPage = appService.listFeaturedAppVOByPage(appQueryRequest);
        return ResultUtils.success(appVOPage);
    }

    /**
     * 【管理员】根据 id 删除任意应用（同时删除关联的对话历史）
     */
    @PostMapping("/delete")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> deleteAppByAdmin(@RequestBody DeleteRequest deleteRequest) {
        ThrowUtils.throwIf(deleteRequest == null || deleteRequest.getId() == null, ErrorCode.PARAMS_ERROR);

        // 查询应用
        App app = appService.getById(deleteRequest.getId());
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR);

        // 先删除关联的对话历史
        boolean chatHistoryDeleted = chatHistoryService.deleteByAppId(deleteRequest.getId());
        if (!chatHistoryDeleted) {
            throw new BusinessException(ErrorCode.OPERATION_ERROR, "删除对话历史失败");
        }

        // 再删除应用
        boolean result = appService.removeById(deleteRequest.getId());
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(true);
    }

    /**
     * 【管理员】根据 id 更新任意应用（支持更新应用名称、应用封面、优先级）
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updateAppByAdmin(@RequestBody AppUpdateRequest appUpdateRequest) {
        ThrowUtils.throwIf(appUpdateRequest == null || appUpdateRequest.getId() == null, ErrorCode.PARAMS_ERROR);

        App app = new App();
        BeanUtil.copyProperties(appUpdateRequest, app);
        app.setEditTime(new Date());

        boolean result = appService.updateById(app);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(true);
    }

    /**
     * 【管理员】分页查询应用列表（支持根据除时间外的任何字段查询，每页数量不限）
     */
    @PostMapping("/list/page")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<App>> listAppByPage(@RequestBody AppQueryRequest appQueryRequest) {
        ThrowUtils.throwIf(appQueryRequest == null, ErrorCode.PARAMS_ERROR);

        long pageNum = appQueryRequest.getPageNum();
        long pageSize = appQueryRequest.getPageSize();

        Page<App> appPage = appService.page(new Page<>(pageNum, pageSize), appService.getQueryWrapper(appQueryRequest));

        return ResultUtils.success(appPage);
    }

    /**
     * 【管理员】根据 id 查看应用详情
     */
    @GetMapping("/get")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<App> getAppById(long id) {
        ThrowUtils.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);

        App app = appService.getById(id);
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR);

        return ResultUtils.success(app);
    }

    /**
     * 【管理员】设置应用为精选/取消精选
     */
    @PostMapping("/set/featured")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> setAppFeatured(@RequestBody AppUpdateRequest appUpdateRequest) {
        ThrowUtils.throwIf(appUpdateRequest == null || appUpdateRequest.getId() == null, ErrorCode.PARAMS_ERROR);

        // 查询应用是否存在
        App oldApp = appService.getById(appUpdateRequest.getId());
        ThrowUtils.throwIf(oldApp == null, ErrorCode.NOT_FOUND_ERROR, "应用不存在");

        // 设置优先级（1为精选，0为非精选）
        App app = new App();
        app.setId(appUpdateRequest.getId());
        app.setPriority(appUpdateRequest.getPriority() != null ? appUpdateRequest.getPriority() : 0);
        app.setEditTime(new Date());

        boolean result = appService.updateById(app);
        ThrowUtils.throwIf(!result, ErrorCode.OPERATION_ERROR);

        return ResultUtils.success(true);
    }

    /**
     * 【管理员】分页查询所有应用列表（按创建时间倒序排列）
     */
    @PostMapping("/list/all/page")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<App>> listAllAppsByPage(@RequestBody AppQueryRequest appQueryRequest) {
        ThrowUtils.throwIf(appQueryRequest == null, ErrorCode.PARAMS_ERROR);

        long pageNum = appQueryRequest.getPageNum();
        long pageSize = appQueryRequest.getPageSize();

        // 创建查询条件，按创建时间倒序排列
        QueryWrapper<App> queryWrapper = new QueryWrapper<>();
        queryWrapper.orderByDesc("createTime");

        Page<App> appPage = appService.page(new Page<>(pageNum, pageSize), queryWrapper);
        return ResultUtils.success(appPage);
    }

    /**
     * 与ai对话生成应用
     *
     * @param appId
     * @param message
     * @param request
     * @return
     */
    @GetMapping(value = "/chat/gen/code", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    //redisson令牌桶限流
    @RateLimit(limitType = RateLimitType.USER, rate = 5, rateInterval = 60, message = "AI 对话请求过于频繁，请稍后再试")
    public Flux<ServerSentEvent<String>> chatToGenCode(@RequestParam Long appId, @RequestParam String message, HttpServletRequest request) {
        ThrowUtils.throwIf(appId == null || appId <= 0, ErrorCode.PARAMS_ERROR, "应用ID无效");
        ThrowUtils.throwIf(StrUtil.isBlank(message), ErrorCode.PARAMS_ERROR, "用户消息不能为空");
        User loginUser = userService.getLoginUser(request);

        // 校验权限：应用所有者、精选应用、或空间成员可以访问
        checkAppAccessPermission(appId, loginUser);

        // IP级别限流检查（每分钟10次）
        ipUtils.checkIpRateLimit(request);

        Flux<String> contentFlux = appService.chatToGenCode(appId, message, loginUser);
        return contentFlux
                .map(chunk -> {
                    Map<String, String> wrapper = Map.of("d", chunk);
                    String jsonData = JSONUtil.toJsonStr(wrapper);
                    return ServerSentEvent.<String>builder()
                            .data(jsonData)
                            .build();
                })
                .concatWith(Mono.just(
                        // 发送结束事件
                        ServerSentEvent.<String>builder()
                                .data("[DONE]")
                                .build()
                ));
    }

    /**
     * 取消正在进行的代码生成
     *
     * @param appId   应用 ID
     * @param request 请求对象
     * @return 取消结果
     */
    @PostMapping("/chat/cancel")
    public BaseResponse<Boolean> cancelCodeGeneration(@RequestParam Long appId,
                                                      HttpServletRequest request) {
        // 参数校验
        ThrowUtils.throwIf(appId == null || appId <= 0, ErrorCode.PARAMS_ERROR, "应用ID无效");
        // 获取当前登录用户
        User loginUser = userService.getLoginUser(request);

        // 执行取消操作
        boolean success = appService.cancelGeneration(appId, loginUser.getId());

        return ResultUtils.success(success);
    }

    /**
     * 部署应用
     *
     * @param appDeployRequest
     * @param request
     * @return
     */
    @PostMapping("/deploy")
    public BaseResponse<String> deployApp(@RequestBody AppDeployRequest appDeployRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(appDeployRequest == null, ErrorCode.PARAMS_ERROR);
        Long appId = appDeployRequest.getAppId();
        ThrowUtils.throwIf(appId == null || appId <= 0, ErrorCode.PARAMS_ERROR, "应用 ID 不能为空");
        User loginUser = userService.getLoginUser(request);
        String deployUrl = appService.deployApp(appId, loginUser);
        return ResultUtils.success(deployUrl);
    }

    /**
     * 查询所有应用类别
     *
     * @return 应用类别列表
     */
    @GetMapping("/types")
    public BaseResponse<List<AppTypeVO>> listAllAppTypes() {
        List<AppTypeVO> appTypeList = Arrays.stream(AppTypeEnum.values()).map(appType -> {
            AppTypeVO appTypeVO = new AppTypeVO();
            appTypeVO.setCode(appType.getCode());
            appTypeVO.setText(appType.getText());
            appTypeVO.setCategory(appType.getCategory());
            appTypeVO.setCategoryName(appType.getCategoryName());
            return appTypeVO;
        }).collect(Collectors.toList());
        return ResultUtils.success(appTypeList);
    }

    /**
     * 获取所有预设提示词列表
     */
    @GetMapping("/preset-prompts")
    public BaseResponse<List<PresetPromptVO>> listAllPresetPrompts() {
        List<PresetPromptVO> promptList = Arrays.stream(PresetPromptEnum.values()).map(prompt -> {
            PresetPromptVO vo = new PresetPromptVO();
            vo.setLabel(prompt.getLabel());
            vo.setPrompt(prompt.getPrompt());
            return vo;
        }).collect(Collectors.toList());
        return ResultUtils.success(promptList);
    }

    /**
     * 检查应用访问权限
     * 应用所有者、精选应用、或空间成员可以访问
     *
     * @param appId     应用ID
     * @param loginUser 当前登录用户
     */
    private void checkAppAccessPermission(Long appId, User loginUser) {
        App app = appService.getById(appId);
        ThrowUtils.throwIf(app == null, ErrorCode.NOT_FOUND_ERROR, "应用不存在");

        // 应用所有者可以访问
        if (app.getUserId().equals(loginUser.getId())) {
            return;
        }

        // 精选应用可以访问
        if (app.getPriority() != null && app.getPriority() == 1) {
            return;
        }

        // 空间成员可以访问
        if (app.getSpaceId() != null) {
            QueryWrapper queryWrapper = new QueryWrapper();
            queryWrapper.eq("spaceId", app.getSpaceId());
            queryWrapper.eq("userId", loginUser.getId());
            boolean isSpaceMember = spaceUserService.count(queryWrapper) > 0;
            if (isSpaceMember) {
                return;
            }
        }

        // 无权限访问
        throw new BusinessException(ErrorCode.NO_AUTH_ERROR, "无权限访问该应用");
    }

}

File: ai-code-antd-frontend/src/services/backend/appController.ts
// @ts-ignore
/* eslint-disable */
import { request } from '@umijs/max';

/**
 * 添加应用
 * 创建一个新的 AI 应用
 * @param body - 应用添加请求参数
 * @param options - 可选的请求配置
 * @returns 应用 ID
 */
export async function addApp(body: API.AppAddRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseLong>('/app/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 聊天生成代码
 * 通过 AI 聊天生成应用代码
 * @param params - 聊天生成代码参数
 * @param options - 可选的请求配置
 * @returns 服务器发送事件字符串数组
 */
export async function chatToGenCode(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.chatToGenCodeParams,
  options?: { [key: string]: any },
) {
  return request<API.ServerSentEventString[]>('/app/chat/gen/code', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/**
 * 管理员删除应用
 * 管理员删除指定的应用
 * @param body - 删除请求参数
 * @param options - 可选的请求配置
 * @returns 删除是否成功
 */
export async function deleteAppByAdmin(body: API.DeleteRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/app/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 用户删除应用
 * 用户删除自己创建的应用
 * @param body - 删除请求参数
 * @param options - 可选的请求配置
 * @returns 删除是否成功
 */
export async function deleteApp(body: API.DeleteRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/app/delete/user', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 部署应用
 * 部署应用到服务器
 * @param body - 应用部署请求参数
 * @param options - 可选的请求配置
 * @returns 部署结果
 */
export async function deployApp(body: API.AppDeployRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseString>('/app/deploy', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 下载应用代码
 * 下载指定应用的源代码
 * @param params - 下载应用代码参数
 * @param options - 可选的请求配置
 * @returns 应用代码文件
 */
export async function downloadAppCode(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.downloadAppCodeParams,
  options?: { [key: string]: any },
) {
  const { appId: param0, ...queryParams } = params;
  return request<any>(`/app/download/${param0}`, {
    method: 'GET',
    params: { ...queryParams },
    ...(options || {}),
  });
}

/**
 * 分页查询精选应用
 * 获取精选应用列表
 * @param body - 应用查询请求参数
 * @param options - 可选的请求配置
 * @returns 分页应用视图对象
 */
export async function listFeaturedAppVoByPage(
  body: API.AppQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageAppVO>('/app/featured/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 根据ID获取应用
 * 获取指定应用的详细信息
 * @param params - 获取应用参数
 * @param options - 可选的请求配置
 * @returns 应用对象
 */
export async function getAppById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAppByIdParams,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseApp>('/app/get', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/**
 * 根据ID获取应用视图
 * 获取指定应用的视图信息
 * @param params - 获取应用视图参数
 * @param options - 可选的请求配置
 * @returns 应用视图对象
 */
export async function getAppVoById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAppVOByIdParams,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseAppVO>('/app/get/vo', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/**
 * 分页查询所有应用
 * 管理员分页获取所有应用
 * @param body - 应用查询请求参数
 * @param options - 可选的请求配置
 * @returns 分页应用对象
 */
export async function listAllAppsByPage(
  body: API.AppQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageApp>('/app/list/all/page', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 分页查询应用
 * 分页获取应用列表
 * @param body - 应用查询请求参数
 * @param options - 可选的请求配置
 * @returns 分页应用对象
 */
export async function listAppByPage(body: API.AppQueryRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponsePageApp>('/app/list/page', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 分页查询我的应用
 * 获取当前用户创建的应用列表
 * @param body - 应用查询请求参数
 * @param options - 可选的请求配置
 * @returns 分页应用视图对象
 */
export async function listMyAppVoByPage(
  body: API.AppQueryRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponsePageAppVO>('/app/my/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 设置应用精选状态
 * 管理员设置应用是否为精选
 * @param body - 应用更新请求参数
 * @param options - 可选的请求配置
 * @returns 设置是否成功
 */
export async function setAppFeatured(body: API.AppUpdateRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/app/set/featured', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 获取所有应用类型
 * 获取应用类型列表
 * @param options - 可选的请求配置
 * @returns 应用类型视图对象列表
 */
export async function listAllAppTypes(options?: { [key: string]: any }) {
  return request<API.BaseResponseListAppTypeVO>('/app/types', {
    method: 'GET',
    ...(options || {}),
  });
}

/**
 * 获取所有预设提示词
 * 获取预设提示词列表
 * @param options - 可选的请求配置
 * @returns 预设提示词视图对象列表
 */
export async function listAllPresetPrompts(options?: { [key: string]: any }) {
  return request<API.BaseResponseListPresetPromptVO>('/app/preset-prompts', {
    method: 'GET',
    ...(options || {}),
  });
}

/**
 * 管理员更新应用
 * 管理员更新应用信息
 * @param body - 应用更新请求参数
 * @param options - 可选的请求配置
 * @returns 更新是否成功
 */
export async function updateAppByAdmin(
  body: API.AppUpdateRequest,
  options?: { [key: string]: any },
) {
  return request<API.BaseResponseBoolean>('/app/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}

/**
 * 用户更新应用
 * 用户更新自己创建的应用信息
 * @param body - 应用更新请求参数
 * @param options - 可选的请求配置
 * @returns 更新是否成功
 */
export async function updateApp(body: API.AppUpdateRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/app/update/user', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  });
}
